pkg/storageincentives/redistributionstate.go

```diff
@@ -0,0 +1,151 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package storageincentives
+
+import (
+	"context"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethersphere/bee/pkg/log"
+	"github.com/ethersphere/bee/pkg/settlement/swap/erc20"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/storageincentives/redistribution"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"math/big"
+	"sync"
+)
+
+const redistributionStatusKey = "redistribution_state"
+const loggerNameNode = "nodestatus"
+
+type RedistributionState struct {
+	stateStore     storage.StateStorer
+	erc20Service   erc20.Service
+	logger         log.Logger
+	overlay        swarm.Address
+	mtx            sync.Mutex
+	status         Status
+	currentBalance *big.Int
+	contract       redistribution.Contract
+}
+
+// Status provide internal status of the nodes in the redistribution game
+type Status struct {
+	Phase           PhaseType
+	IsFrozen        bool
+	Round           uint64
+	LastWonRound    uint64
+	LastPlayedRound uint64
+	LastFrozenRound uint64
+	Block           uint64
+	Reward          *big.Int
+	Fees            *big.Int
+}
+
+func NewRedistributionState(logger log.Logger, stateStore storage.StateStorer, erc20Service erc20.Service, contract redistribution.Contract) RedistributionState {
+	return RedistributionState{
+		stateStore:     stateStore,
+		erc20Service:   erc20Service,
+		logger:         logger.WithName(loggerNameNode).Register(),
+		currentBalance: big.NewInt(0),
+		contract:       contract,
+		status: Status{
+			Round:  0,
+			Block:  0,
+			Reward: nil,
+			Fees:   big.NewInt(0),
+		},
+	}
+}
+
+func (n *RedistributionState) SetCurrentEvent(p PhaseType, r uint64, b uint64) {
+	n.mtx.Lock()
+	defer n.mtx.Unlock()
+	n.status.Phase = p
+	n.status.Round = r
+	n.status.Block = b
+	n.saveStatus()
+}
+
+func (n *RedistributionState) IsFrozen(f bool) {
+	n.mtx.Lock()
+	defer n.mtx.Unlock()
+	n.status.IsFrozen = f
+	n.saveStatus()
+}
+
+func (n *RedistributionState) SetLastWonRound(r uint64) {
+	n.mtx.Lock()
+	defer n.mtx.Unlock()
+	n.status.LastWonRound = r
+	n.saveStatus()
+}
+
+func (n *RedistributionState) SetLastFrozenRound(f uint64) {
+	n.mtx.Lock()
+	defer n.mtx.Unlock()
+	n.status.LastFrozenRound = f
+	n.saveStatus()
+}
+
+func (n *RedistributionState) SetLastPlayedRound(p uint64) {
+	n.mtx.Lock()
+	defer n.mtx.Unlock()
+	n.status.LastPlayedRound = p
+	n.saveStatus()
+}
+
+// AddFee sets the internal node status
+func (n *RedistributionState) AddFee(ctx context.Context, txHash common.Hash) {
+	n.mtx.Lock()
+	defer n.mtx.Unlock()
+	n.status.Fees.Add(n.status.Fees, n.contract.Fee(ctx, txHash))
+	n.saveStatus()
+}
+
+// CalculateWinnerReward calculates the reward for the winner
+func (n *RedistributionState) CalculateWinnerReward(ctx context.Context) error {
+	n.mtx.Lock()
+	defer n.mtx.Unlock()
+	currentBalance, err := n.erc20Service.BalanceOf(ctx, common.HexToAddress(n.overlay.String()))
```

Comment:
Mutex in `CalculateWinnerReward` seem to be overlocking. 

Lock is acquired on entire method which would include call to external service `n.erc20Service.BalanceOf` in same lock. If this method call takes too much time then all call to all other methods of `RedistributionState` would block further propagating blocking to other logic.

Could we reduce locking of `CalculateWinnerReward` logic down to the smallest as possible?

Reply:
So we could only apply lock to this statement instead of locking all method

`n.status.Reward.Add(n.status.Reward, currentBalance.Sub(currentBalance, n.currentBalance))`